### ISSUE TITLE
Parsing a qobj with bfunc raises an error

### ISSUE THREAD

**Issue by kdk** [Qiskit/qiskit-aer#49]
### Informations

- **Qiskit AER version**: 0.1.1
- **Operating system**: MacOS

### What is the current behavior?

In testing https://github.com/Qiskit/qiskit-terra/pull/1778 with the Aer provider, submitting a conditional circuit with a qobj-compliant conditional raises a `AerBackend: simulation failed` error from https://github.com/Qiskit/qiskit-aer/blob/dac153d6730e4de9d7787b0de8540855a26b81e6/src/base/controller.hpp#L284 . Printing out the exception yields:
```
{'job_id': None,
'date': None,
'backend_name': None,
'qobj_id': None,
'status': 'ERROR: Failed to load qobj: [json.exception.type_error.302] type must be array, but is number',
'backend_version': None,
'success': False}
```

### Steps to reproduce the problem
Checkout https://github.com/Qiskit/qiskit-terra/pull/1778 and run any conditional circuit. e.g.
```
qr = qk.QuantumRegister(2)
cr1 = qk.ClassicalRegister(1)
cr2 = qk.ClassicalRegister(1)
qc = qk.QuantumCircuit(qr, cr1, cr2)

qc.h(qr[0])
qc.measure(qr[0],cr1)
qc.x(qr[1]).c_if(cr1, 1)
qc.measure(qr[1],cr2)
```
The resulting qobj
```
{'config': {'memory_slots': 2, 'n_qubits': 2},
 'experiments': [{'config': {'memory_slots': 2, 'n_qubits': 2},
                  'header': {'clbit_labels': [['c0', 0], ['c1', 0]],
                             'creg_sizes': [['c0', 1], ['c1', 1]],
                             'memory_slots': 2,
                             'n_qubits': 2,
                             'name': 'circuit0',
                             'qreg_sizes': [['q0', 2]],
                             'qubit_labels': [['q0', 0], ['q0', 1]]},
                  'instructions': [{'name': 'h', 'qubits': [0]},
                                   {'memory': [0],
                                    'name': 'measure',
                                    'qubits': [0],
                                    'register': [0]},
                                   {'mask': '0x1',
                                    'name': 'bfunc',
                                    'register': 2,
                                    'relation': '==',
                                    'val': '0x1'},
                                   {'conditional': 2,
                                    'name': 'x',
                                    'qubits': [1]},
                                   {'memory': [1],
                                    'name': 'measure',
                                    'qubits': [1],
                                    'register': [1]}]}],
 'header': {},
 'qobj_id': '94b4fbcc-e12e-47bc-898c-4106b4686867',
 'schema_version': '1.0.0',
 'type': 'QASM'}
```
contains a `bfunc` with `register: 2`, which is causing the JSON parser to throw.

### What is the expected behavior?

It looks like the JSON->op parser ( https://github.com/Qiskit/qiskit-aer/blob/master/src/framework/operations.hpp#L474 ) expects memory and register to be arrays ( https://github.com/Qiskit/qiskit-aer/blob/master/src/framework/operations.hpp#L38 ) but in the qobj-spec, `bfunc` instructions are a special case where they are singleton integers.

**Comment by chriseclectic** [comment]
Can you test if with this change in `json_to_op_bfunc` in `operations.hpp` fixes it?

```cpp
JSON::get_value(op.memory, "memory", js);
JSON::get_value(op.registers, "register", js);
```

to

```cpp
uint_t tmp;
if (JSON::get_value(tmp, "register", js)) {
  op.registers.push_back(tmp);
}
if (JSON::get_value(tmp, "memory", js)) {
  op.memory.push_back(tmp);
}
```

NB: Do we actually need the `"memory"` part for bfunc? If only register used by the spec then the last part can just be removed.

**Comment by kdk** [comment]
That gets it past the JSON validation failure and the simulation completes, but the counts returned aren't what I wouldn't expect.

For the circuit above:
```
         ┌───┐┌─┐          
q1_0: |0>┤ H ├┤M├──────────
         └───┘└╥┘┌─────┐┌─┐
q1_1: |0>──────╫─┤  X  ├┤M├
               ║ ├──┴──┤└╥┘
 c2_0: 0 ══════╩═╡ = 1 ╞═╬═
                 └─────┘ ║ 
 c3_0: 0 ════════════════╩═
```
I get (with qobj-conditionals from terra and the fix above):
```
{'0 0': 513, '0 1': 511}
```
With old-style conditionals from terra and master aer I get:
```
{'0 0': 519, '1 1': 505}
```

Looking briefly, it looks like `apply_bfunc` is making the right comparison, but its result is not being written to the provided register. (I'm not at all familiar with the simulator code, but it looks like `apply_bfunc` is returning the boolean result of the comparison, but `State::apply_ops` is not getting that return value written to the register.)

Re: `"memory"`, the spec lists it as "(optional): Memory slot in which to store the boolean function result" presumably for debugging.

**Comment by chriseclectic** [comment]
@kdk ok thanks for looking into it. I've tried fixing it in PR #55 if you want to test it.

Note: there might still be bugs in the actual logic of the bfunc function since it hasn't really well tested.

**Comment by kdk** [comment]
:+1: Awesome, tested on #55 and it looks good. Thanks!

**Comment by chriseclectic** [comment]
Closed by #55
